packages/vite/src/node/server/middlewares/spaFallback.ts

```diff
@@ -0,0 +1,30 @@
+import fs from 'fs'
+import history from 'connect-history-api-fallback'
+import path from 'path'
+import { Connect } from 'types/connect'
+import { createDebugger } from '../../utils'
+
+export function spaFallbackMiddleware(
+  root: string
+): Connect.NextHandleFunction {
+  // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+  return function viteSpaFallbackMiddleware(req, res, next) {
+    return history({
+      logger: createDebugger('vite:spa-fallback'),
+      // support /dir/ without explicit index.html
+      rewrites: [
+        {
+          from: /\/$/,
+          to({ parsedUrl }: any) {
+            const rewritten = parsedUrl.pathname + 'index.html'
+            if (fs.existsSync(path.join(root, rewritten))) {
+              return rewritten
+            } else {
+              return `/index.html`
+            }
+          }
+        }
+      ]
+    })(req, res, next)
+  }
```

Comment:
Just returning `history()` would create a new spa fallback middleware in each request. I think we need to create it beforehand and wrap it in the named function, no? Something like:

```suggestion
  const spaFallbackMiddleware = return history({
    logger: createDebugger('vite:spa-fallback'),
    // support /dir/ without explicit index.html
    rewrites: [
      {
        from: /\/$/,
        to({ parsedUrl }: any) {
          const rewritten = parsedUrl.pathname + 'index.html'
          if (fs.existsSync(path.join(root, rewritten))) {
            return rewritten
          } else {
            return `/index.html`
          }
        }
      }
    ]
  })
  
  // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
  return function viteSpaFallbackMiddleware(req, res, next) {
    return spaFallbackMiddleware(req, res, next)
  }
```